Migrate /chatters commands to use Helix api

cbclemmer opened the pull request:
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description
closes #4019

## Features:
- Migrates /chatters command will use helix api or cache
	- If the user is neither mod or broadcaster the list is taken from the user list cache
	- If the user has mod rights the list is taken from the chat/chatters helix api command
- The old irc commands for /chatters have been removed based on comment in issue

<!-- If applicable, please include a summary of what you've changed and what issue is fixed. In the case of a bug fix, please include steps to reproduce the bug so the pull request can be tested -->

pajlada reviewed (commented):
Thanks for working on this!

These changes are both pretty heavy, and while they rely on the same underlying functionality (the cursor-pages-fetching thing), it would be a lot easier to review if I could focus on a single functionality rather than two, since most feedback I have would apply to both.

So in short, could you undo the changes either to the /mods command or to the /chatters command (I have no preference) in this PR?

After that's done, I will go ahead and review the code and come with more proper/full feedback

cbclemmer commented:
@pajlada Code and pr message has been updated. I've got everything in another branch to merge in later. There were a few functions that will be shared for /chatters and /mods so some functions look like they shouldn't be their own thing, but they'll be useful later.

pajlada commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -621,7 +592,9 @@ const QString &TwitchChannel::popoutPlayerUrl()
 int TwitchChannel::chatterCount()
 {
-    return this->chatterCount_;
+    return this->hasModRights()
+        ? this->chatterCount_
+        : this->accessChatters()->getNumChatters();
```
These two numbers are very different and used for different purposes, portraying them both as the same number is in my opinion confusing.

cbclemmer replied:
`chatterCount_` has been removed because it cannot be used by non mods anyways. I've simplified this function down to just using `this->accessChatters()->getNumChatters()`

pajlada commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -135,10 +111,6 @@ TwitchChannel::TwitchChannel(const QString &name)
         }
     });
 
-    // timers
-    QObject::connect(&this->chattersListTimer_, &QTimer::timeout, [=] {
-        this->refreshChatters();
-    });
```
Why do we want to disable this timer?

cbclemmer replied:
The chatter list can only be refreshed if you are a moderator now and it is automatically refreshed if a mod uses the /chatters command. So the only reason to update it is so that mods would have an updated @ list. I don't mind adding it back but I didn't know if it is worth it at that point.

pajlada commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -164,7 +136,6 @@ TwitchChannel::TwitchChannel(const QString &name)
 void TwitchChannel::initialize()
 {
     this->fetchDisplayName();
-    this->refreshChatters();
```
Why do we no longer want to refresh chatters on startup?

cbclemmer replied:
Same as the timer essentially. If we need to make sure that mods have an updated chatter list it can be reverted.

cbclemmer replied:
Mods should probably always have access to an accurate @ list. Reverted code.

pajlada commented on src/controllers/commands/CommandController.cpp:
```diff
@@ -889,9 +895,27 @@ void CommandController::initialize(Settings &, Paths &paths)
+
+            // Fetch from chatter cache when user does not have mod rights
+            if (!twitchChannel->hasModRights()) {
+                addChatterCountMessage();
```
Given my previous comment above, I would prefer to return an error message for non-mods rather than providing inaccurate information.

cbclemmer replied:
Added an error message.

pajlada commented on src/controllers/commands/CommandController.cpp:
```diff
@@ -889,9 +895,27 @@ void CommandController::initialize(Settings &, Paths &paths)
+            }
+
+            // Use helix api when user has mod rights
+            twitchChannel->refreshChatters(
```
This will end up making N requests fetching all viewers, it will probably be best to also provide a non-paginated version that this command can use directly

The way I envision it working is the user types `/chatters`, we make a request using `getHelix()->getChatters(channelsomething, /* how many users to fetch */ 1, /* pagination cursor */ boost::none, successCB, failCB)` where successCB only reads the total field and discards the rest of the payload

cbclemmer replied:
Good point, I forgot there is a total in the first returned request. I'll make a separate function to just get the total. The `Helix::GetChatters` function is still relevant because we need it for refreshing the @ list.

pajlada commented on src/providers/twitch/api/Helix.hpp:
```diff
@@ -329,6 +330,16 @@ struct HelixChatSettings {
+class HelixUserList {
+    public:
+        std::unordered_set<QString> users;
+        int total;
```
Preferably `total` isn't included in this data structure to keep it as small/to-the-point as possible

cbclemmer replied:
Ok, I don't think I need this class then, I'll just pass back the unordered set.

pajlada replied:
Yeah that works well :+1:

pajlada commented on src/providers/twitch/api/Helix.hpp:
```diff
@@ -329,6 +330,16 @@ struct HelixChatSettings {
+
+        HelixUserList();
+
+        void AddUsersFromResponse(const QJsonObject &response);
```
Casing should be `addUsersFromResponse`

Preferably the function is protected/private with only the Helix class having access to it so users don't get the idea that it can be called

cbclemmer replied:
Class deleted

pajlada commented on src/providers/twitch/api/Helix.hpp:
```diff
@@ -329,6 +330,16 @@ struct HelixChatSettings {
+        std::unordered_set<QString> users;
+        int total;
+
+        HelixUserList();
```
The constructor doesn't do anything special so it can be removed entirely

cbclemmer replied:
Class deleted

pajlada commented on src/providers/twitch/api/Helix.cpp:
```diff
@@ -8,6 +8,19 @@
+        auto user = jsonStream.toObject().value("user_login").toString();
+        users.insert(user);
+    }
+    users = std::move(users);
```
This line doesn't do anything

cbclemmer replied:
class deleted

pajlada commented on src/providers/twitch/api/Helix.cpp:
```diff
@@ -8,6 +8,19 @@
+    for (const auto jsonStream : response.value("data").toArray()) 
+    {
+        auto user = jsonStream.toObject().value("user_login").toString();
+        users.insert(user);
```
Use `this->` when accessing member variable

so `this->users.insert(user);` instead of `users.insert(user);`

cbclemmer replied:
class deleted

pajlada commented on src/providers/twitch/api/Helix.cpp:
```diff
@@ -1991,6 +2004,157 @@ void Helix::sendWhisper(
+
+// Recursive function with a maximun page of 50
+void Helix::getApiListRecursive(
+    HelixUserList *list,
```
Let's avoid a pointer here. You could pass it by reference, pass a `shared_ptr`, or move it from call to call.

cbclemmer replied:
Moved reference to list to the parent function referenced in a callback.

pajlada commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -621,7 +597,7 @@ const QString &TwitchChannel::popoutPlayerUrl()
 int TwitchChannel::chatterCount()
 {
-    return this->chatterCount_;
+    return this->accessChatters()->getNumChatters();
```
We still can't use `getNumChatters` here since it relies on the chatters that are part of the `ChatterSet` and the `ChatterSet` a) doesn't remove stale entries b) doesn't go beyond 2000. So keeping and using the `chatterCount_` for this specific purpose still makes sense imo

cbclemmer replied:
It now sets `chatterCount_` whenever `refreshChatters` is run and this function now uses `chatterCount_` again.

pajlada commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -917,31 +898,16 @@ void TwitchChannel::refreshChatters()
+    getHelix()->getChatters(
+        this->roomId(),
+        getApp()->accounts->twitch.getCurrent()->getUserId(),
+        [this](HelixUserList *chatterList) {
```
We need to use the same `weakOf` 'hack' that the previous TMI version used to ensure the callback doesn't attempt to use a TwitchChannel that has been deleted

cbclemmer replied:
Not sure what you mean here. I tried to implement it based on the other functions and I think I got it...

cbclemmer commented:
@pajlada I changed the helper function for helix to be a more generalized version of `makeRequest` that can be paginated or not. I'll use this for the /mods pr too. It now can pass in a function that does what it needs to the response from the api and can recursively loop until the end of the list. The /chatters command now just gets one page of chat/chatters api and returns the the total property value. I also changed the error type to be a general error because there isn't any special error types with these requests.

cbclemmer commented on src/providers/twitch/api/Helix.hpp:
```diff
@@ -519,6 +520,15 @@ enum class HelixWhisperError {  // /w
     Forwarded,
 };  // /w
 
+enum class HelixGeneralError {
+    Unknown,
+    UserMissingScope,
+    UserNotAuthorized,
+
+    // The error message is forwarded directly from the Twitch API
+    Forwarded,
+};
```
Ah, ok that makes sense.

Felanbird commented on src/controllers/commands/CommandController.cpp:
```diff
@@ -878,23 +878,79 @@ void CommandController::initialize(Settings &, Paths &paths)
+            return errorMessage;
+        };
 
+        if (words.size() != 1)
+        {
+            channel->addMessage(makeSystemMessage("Usage: /chatters"));
+            return "";
+        }
```
Why do we care about the input for this command?

cbclemmer replied:
I was just following the convention of the other commands, but I guess those were using arguments and could lead being misused. I don't mind removing it.

cbclemmer replied:
nvm, it was already removed

pajlada commented:
@cbclemmer uno reverse card - I committed some stuff to your PR that I'd like you to review/approve. Most commits are small and do very little, they'd be easy to go through and it's easy to understand why. However, I made a big refactor PR for the pagination logic to change the pattern we use. This changes the pattern for paginated requests in this PR and for future paginated requests to need two functions; 1 public one that users (e.g. `CommandController` or `TwitchChannel`) can use which has a simple, single-use API and 1 private one that the public function will use & recursively call.
The big refactor commit can be found [here `a3aed80` (#4088)](https://github.com/Chatterino/chatterino2/pull/4088/commits/a3aed8084f02e110745de9a008c89a77c803d427) and has a commit message where I try to explain all of the changes I made and why.

Let me know if you have any questions!

cbclemmer commented:
@pajlada That all looks good. The reason I wanted to try to make a more general request was because the handling for the status codes and resulting errors were largely copied and pasted so I tried to make a framework to reduce that for the stuff I was writing. It does make a lot more sense to refactor everything later when everything is using a standardized way of handling errors like you commented before. I'm mainly a C# dev and garbage collection is usually an afterthought for me because of that, so thanks for ironing that out. Looks good to me 👍

Felanbird reviewed (approved):
I'm not 100% on the fact that we are deprecating the command very early, but `/chatters` usage is probably more common among stable users, so we could take another look at this around Feb when the deprecation of all other commands kicks in.
